sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/AsyncCacheNonBlocking.java

```diff
@@ -0,0 +1,225 @@
+package com.azure.cosmos.implementation.caches;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Mono;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Function;
+
+/**
+ *
+ * @param <TKey>
+ * @param <TValue>
+ *
+ * This is a thread safe AsyncCache that allows refreshing values in the background.
+ * The benefits of AsyncCacheNonBlocking over AsyncCache is it keeps stale values until the refresh is completed.
+ * AsyncCache removes values causing it to block all requests until the refresh is complete.
+ * 1. For example 1 replica moved out of the 4 replicas available. 3 replicas could still be processing requests.
+ *    The request going to the 1 stale replica would be retried.
+ * 2. AsyncCacheNonBlocking updates the value in the cache rather than recreating it on each refresh. This will help reduce object creation.
+ *
+ */
+public class AsyncCacheNonBlocking<TKey, TValue> {
+    private final static Logger logger = LoggerFactory.getLogger(AsyncCacheNonBlocking.class);
+    private final ConcurrentHashMap<TKey, AsyncLazyWithRefresh<TValue>> values;
+    private final IEqualityComparer<TKey> keyEqualityComparer;
+
+    public AsyncCacheNonBlocking(ConcurrentHashMap<TKey, AsyncLazyWithRefresh<TValue>> values, IEqualityComparer<TKey> keyEqualityComparer) {
+        this.values = values;
+        this.keyEqualityComparer = keyEqualityComparer;
+    }
+
+    public AsyncCacheNonBlocking(IEqualityComparer<TKey> keyEqualityComparer) {
+        this(new ConcurrentHashMap<>(), keyEqualityComparer);
+    }
+
+    public AsyncCacheNonBlocking() {
+        this((value1, value2) -> {
+            if (value1 == value2)
+                return true;
+            if (value1 == null || value2 == null)
+                return false;
+            return value1.equals(value2);
+        });
+    }
+
+    /**
+     *
+     * <p>
+     * If another initialization function is already running, new initialization function will not be started.
+     * The result will be result of currently running initialization function.
+     * </p>
+     *
+     * <p>
+     * If previous initialization function is successfully completed it will return the value. It is possible this
+     * value is stale and will only be updated after the force refresh task is complete.
+     * </p>
+     *
+     * <p>
+     * If previous initialization function failed - new one will be launched.
+     * </p>
+     *
+     * @param key Key for which to get a value.
+     * @param singleValueInitFunc Initialization function.
+     * @param forceRefresh Force refresh for refreshing the cache
+     * @return Cached value or value returned by initialization function.
+     */
+    public Mono<TValue> getAsync(
+        TKey key,
+        Function<TValue, Mono<TValue>> singleValueInitFunc,
+        boolean forceRefresh) {
```

Comment:
the forceRefresh should be determined by the currentValue vs obsoleteValue. 

https://github.com/Azure/azure-cosmos-dotnet-v3/blob/4875e4572ccc072320ae52c030c42d92480c9a56/Microsoft.Azure.Cosmos/src/Routing/AsyncCacheNonBlocking.cs#L86

Reply:
Yes, it should be determined by the currentValue vs obsoleteValue. Updated the code logic.